test/test_collision_resolution.py

```diff
@@ -37,49 +39,181 @@ def f(key):
+            jnp.linalg.norm(penetration_after)
+            < 1e-3 * jnp.linalg.norm(penetration_before),
         )
-        # wither the collision was resolved, or the balls are moving apart
+        # either the collision was resolved, or the balls are moving apart
         return (
-            v1_away
-            & v2_away
-            & res_first_collision
-            & (jnp.logical_xor(no_collision, moving_apart))
+            velocities_away,
+            res_first_collision,
+            jnp.logical_xor(no_collision, moving_apart),
+            jnp.array([p1, p2, v1, v2, [r1, r2], penetration_before, [dist, dist]]),
+            jnp.array(
+                [
+                    body1.position,
+                    body2.position,
+                    body1.velocity,
+                    body2.velocity,
+                    penetration_after,
+                ]
+            ),
         )
 
     N = 1000
     f = jax.vmap(f)
-    out = f(jr.split(key, N))
+    (
+        velocities_away,
+        was_collision,
+        collision_resolved_xor_didnt_have_to_be,
+        start_info,
+        end_info,
+    ) = f(jr.split(key, N))
+    all_conditions = (
+        velocities_away & was_collision & collision_resolved_xor_didnt_have_to_be
+    )
+    start_useful = start_info[~all_conditions]
+    end_useful = end_info[~all_conditions]
+    if end_useful.shape[0] != 0:
+        jnp.set_printoptions(suppress=True)
+        print(f"\ntotal wrong {end_useful.shape[0] / N}")
+    for i in range(end_useful.shape[0]):
+        print()
+        print(f"start position 1: {start_useful[i, 0]}")
+        print(f"start position 2: {start_useful[i, 1]}")
+        print(f"end position 1: {end_useful[i, 0]}")
+        print(f"end position 2: {end_useful[i, 1]}")
+        print(f"start velocity 1: {start_useful[i, 2]}")
+        print(f"start velocity 2: {start_useful[i, 3]}")
+        print(f"end velocity 1: {end_useful[i, 2]}")
+        print(f"end velocity 2: {end_useful[i, 3]}")
+        print(f"radius 1: {start_useful[i, 4][0]}")
+        print(f"radius 2: {start_useful[i, 4][1]}")
+        print(f"distance: {start_useful[i, 6][0]}")
+        print(f"end position difference: {end_useful[i, 0] - end_useful[i, 1]}")
+        print(f"end velocity difference: {end_useful[i, 2] - end_useful[i, 3]}")
+        print(f"penetration before: {start_useful[i, 5]}")
+        print(f"penetration after: {end_useful[i, 4]}")
+
+    assert jnp.all(was_collision), "there wasnt a collision"
+    assert jnp.all(
+        collision_resolved_xor_didnt_have_to_be
+    ), "collision was not resolved, or was resolved unnecessarily"
+    assert jnp.all(velocities_away), "velocities arent away"
+
+
+def test_triangle_circle_angular():
+    # todo: so this test is not passing,
+    #  because global support of the triangle along epa vector is [0.8 3.5],
+    #  but i want it to be [1.22, 2.63] (not in a vertex)
```

Comment:
evidently, this test doesn't pass, and I want it to pass, so before I implement something stupid, I'd like to ask: given the existing GJK and EPA algorithms, is there an easy way to get a good approximation for the contact point (avgCollisionPoint)?

the problem occurs when not using circles, as the global support returns a vertex, whereas for 
```
contact_point = (
        body1.shape.get_global_support(-unit_collision_vector)
        + body2.shape.get_global_support(unit_collision_vector)
    ) / 2  # in _resolve_collision()
```
I would want points on the edge (global support then is a wrong function of course, as support in a direction of perpendicular to an edge can be any point on the edge) where the EPA vector can be applied.

if there is no easy way currently, I thought of the following rules: 
1. either a vertex is penetrating, or one shape is a circle
2. in both scenarios, I need a perpendicular to the edge from a vertex or a point on the circumference.
3. if two circles, or 2+ vertices are penetrating (at least one from each body), take average

However, properly implementing that will take some time and increase the compute cost for collision resolution.

Here is the visualization of the test:
<img width="661" alt="image" src="https://github.com/DelftMercurians/cotix/assets/27450370/c13cf25b-79aa-444b-9ddc-33461067a1e4">

PS: after a brief look at RSS, I do not think this is implemented correctly there, as the collision point is take as the average of `Aw.getFarthestPointInDirection(penetrationVector)`, so same as currently in our code, since getFarthestPointInDirection returns a vertex for a polygon.

Reply:
also, collision simplex is 
```
[[nan nan]
 [nan nan]
 [nan nan]]
```
on that test, although collision occurs (not sure if that should be the case)

Reply:
oh yeah, the question is to @knyazer of course

Reply:
and here is a geogebra export if you feel like it
[geogebra-export.txt](https://github.com/DelftMercurians/cotix/files/13213421/geogebra-export.txt)

Reply:
Probs you can just assume that somebody is going to pass a contact point to you?

I am thinking about the problem, and while we can extract the actual point via doing some magic, I am not necessarily sure it's correct.

And of course we want the physics solver to be robust to the choice of the collision detection algorithm

Reply:
So where are we gonna get this point from then? And is the simplex correct?

Reply:
simplex is nan is usually when there is no collision detected, or there is a bug in the EPA/GJK :) most likely though this is the reason: EPA cannot really work well when gjk fails, and in this case something obviously went wrong. I don't have an explicit check in EPA (which I probably should add) that detects simplex being nan/ill-formed.

I am not really sure what is the issue, and don't have time to look at it until thursday. Hence you can solve this issue just by passing a contact object to the collision resolution method, where this Contact is custom defined eqx.Module with whatever you want to know, and on thursday I will make it so that it gets all this stuff somehow, I guess.

The problem is related to the fact that get_support function is not defining minkowski difference, it is defining set of points, convex hull of which is going to give the minkowski difference. I quickly looked through the code, and this is a relatively easily fixable bug with EPA giving incorrect result (of course, assuming GJK worked correctly). Effectively instead of taking the last support point, one should compute the closest point to the origin in the whole polygon. I think I even meant to do this at some point, since the current implementation allows to catch bugs much easier (since if anything thorough EPA execution fails, the support is going to be in the wrong direction, so instead of 'almost correct' answer we will get 'completely incorrect'), but I have forgotten.